Teach Integer Range Optimization about ToLength

justinmichaud opened the pull request:
#### d37295c826a2729b3958abacfcfb17ceb11ccd5d
<pre>
Teach Integer Range Optimization about ToLength
<a href="https://bugs.webkit.org/show_bug.cgi?id=302759">https://bugs.webkit.org/show_bug.cgi?id=302759</a>

Reviewed by Keith Miller.

This change allows IRO to infer that ToLength(l) == l if we can prove that
l &gt;= 0.

* Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:

Canonical link: <a href="https://commits.webkit.org/303358@main">https://commits.webkit.org/303358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34d30cbd9c048d817be761b5d69e3dca9515ec63

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131972 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4464 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139485 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83870 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4406 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4226 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100875 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68263 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3136 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118190 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81666 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3030 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/889 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82704 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124035 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111861 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142130 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130479 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4134 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36894 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109249 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4215 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109419 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27750 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3127 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114470 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57355 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4187 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32867 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163446 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4019 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67634 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42494 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4279 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4147 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
```diff
@@ -1641,6 +1641,32 @@ class IntegerRangeOptimizationPhase : public Phase {
+                break;
+            // Consider b = ToLength(a)
+            // If a < 0, then b == 0; b >= a
+            // If a >=0, then b = a;
```
nit `a >= 0`

kmiller68 commented on Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
```diff
@@ -1641,6 +1641,32 @@ class IntegerRangeOptimizationPhase : public Phase {
+                    if (relationship.right() == m_zero
+                        && relationship.offset() >= -1
+                        && relationship.kind() == Relationship::GreaterThan) {
+                        setRelationship(Relationship(node, node->child1().node(), Relationship::Equal, 0));
```
Nit: You don't need the offset here.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
```diff
@@ -1641,6 +1641,32 @@ class IntegerRangeOptimizationPhase : public Phase {
+                    // a > -1
+                    if (relationship.right() == m_zero
+                        && relationship.offset() >= -1
+                        && relationship.kind() == Relationship::GreaterThan) {
```
It feels like there should be a helper to ask questions like this. e.g.
```
bool provablyGreaterThan(Node*, int64_t);
bool provablyGreaterThanEq(Node*, int64_t);
...
```

justinmichaud replied:
I like this idea, I'll add it

kmiller68 commented on Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:
```diff
@@ -1641,6 +1641,32 @@ class IntegerRangeOptimizationPhase : public Phase {
+            if (node->child1().useKind() == UntypedUse)
+                break;
+            // Consider b = ToLength(a)
+            // If a < 0, then b == 0; b >= a
```
It seems like it would be reasonable to add the `b == 0` relationship too, which would be easy if we had the helpers below.

justinmichaud replied:
I believe this would be incorrect, since it won't get pruned. If a was a shadow, then it would break. This sounds like a fun test to write tbh

kmiller68 reviewed (approved):
r=me with nits.

justinmichaud committed:
Teach Integer Range Optimization about ToLength
https://bugs.webkit.org/show_bug.cgi?id=302759

Reviewed by Keith Miller.

This change allows IRO to infer that ToLength(l) == l if we can prove that
l >= 0.

* Source/JavaScriptCore/dfg/DFGIntegerRangeOptimizationPhase.cpp:

Canonical link: https://commits.webkit.org/303358@main
